compiler/rustc_metadata/src/rmeta/encoder.rs

```diff
@@ -1784,16 +1785,23 @@ impl<'a, 'tcx> EncodeContext<'a, 'tcx> {
     }
 
     /// Encodes an index, mapping each trait to its (local) implementations.
-    fn encode_impls(&mut self) -> Lazy<[TraitImpls]> {
-        empty_proc_macro!(self);
-        debug!("EncodeContext::encode_impls()");
+    fn encode_traits_and_impls(&mut self) -> (Lazy<[DefIndex]>, Lazy<[TraitImpls]>) {
+        if self.is_proc_macro {
+            return (Lazy::empty(), Lazy::empty());
+        }
+        debug!("EncodeContext::encode_traits_and_impls()");
         let tcx = self.tcx;
-        let mut visitor = ImplVisitor { tcx, impls: FxHashMap::default() };
+        let mut visitor =
+            TraitsAndImplsVisitor { tcx, impls: FxHashMap::default(), traits: Default::default() };
         tcx.hir().visit_all_item_likes(&mut visitor);
 
+        let mut all_traits = visitor.traits;
```

Comment:
Can this be implemented using `tcx.trait_in_crate(LOCAL_CRATE)`?

Reply:
There's some performance trade-off here.

From performance point of view all the stuff in `fn encode_crate_root` should be encoded by walking the crate once, collecting everything, and then encoding everything, instead of walking the crate many times by `LOCAL_CRATE` queries collecting their specific pieces of data.
But the code would certainly be cleaner and with less duplication with the queries.

I can make the refactoring you suggest later in a separate PR, and we'll land it if doesn't cause observable changes in benchmarks.